Week1 [STEP 3] hogang

hogang1223 opened the pull request:
@Kioding 

키오 안녕하세요!!
2주차가 된 지금에서야 WEEK1의 STEP 3을 제출합니다! ㅎ...

- git
먼저 조언해주신것처럼 커밋 템플릿 만들어서 써봤습니다.
커밋 템플릿은 처음 써보는데... 앞으로 커밋의 문구가 규칙적이고 보기 더 편해질 것 같습니다. 😊

- code
이번 빼빼로 그리기 과제는... 빼빼로 몸통 그릴 때 요구사항 대로 하기위해 조금 머리를 싸맸습니다...
몸통을 먼저 만들고 몸통 길이에 따라 출력해야한다는게 따로따로 함수를 만드려니
토핑을 막대길이에 따라 다른 위치에 뽑아내야되는데... 길이 없이 어떻게 그걸 할 수 있지?
하면서 직접 손으로 빼빼로를 그려보다, 아 두번째줄까지 아예 만들면 되는구나 라고 깨달았습니다 ㅎ... <br>
빼빼로 타입만 입력하면 출력할 수 있게 하고싶었습니다.
처음에 타입을 int값으로 주었다가, 예외처리 할 생각에 머리가 아파져와
애초에 타입에 제한을 주자 라고 생각해서 빼빼로 타입을 enum으로 선언하여,
선언한 빼빼로 타입 외에는 입력할 수 없도록 했습니다.

hogang1223 committed:
feat : STEP3 - 빼빼로 그리기

- .gitmessage.txt 템플릿 작성 및 등록
- STEP3 빼빼로 그리기 코드 작성

Kioding commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -8,6 +8,16 @@
+
 // MARK: Step2: 아이스크림 그리기
 
 // 일단 보자마자 내가 생각한대로
```
`main.swift`에 step2에 사용했던 코드가 들어있네요.
현재 쓰지 않는 코드라면 어떻게 관리하는 게 좋을까요?!

hogang1223 replied:
따로 파일을 새로 생성해서 이동시키는게 좋을까요??

Kioding replied:
`main.swift`는 code의 entry point이기 때문에 말씀해주신 것처럼 파일을 따로 생성하여 관리하면 좋겠네요 👍 

`+`  나중에 branch로 관리해볼 수도 있을 것 같네요!

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,92 @@
+
+// 요구사항 1-1 : 길이에 따른 막대(빼빼로 손잡이)를 그리는 함수
+func printPeperoBar(barLength:Int){
+    for i in 1...barLength{
```
`i`는 무슨 뜻일까요?

Kioding replied:
`(barLength:Int){` 를 붙인 이유가 있을까요? 컨벤션에 맞게 짜려면 어떻게 하는 게 괜찮을까요?? 👀

hogang1223 replied:
핫... 또.. 암 생각없이 빠르게 치다가 생긴....... ㅠ

하.. 진짜 코드 짠 다음에 꼭 재확인 하는 버릇을 가져야겠어요....
그냥 결과만 보고 결과 잘 나오니까 오케오케 하지 않고
다시 한번 한줄 한줄 읽으면서
나도 남도 모두가 보기 편하고 이해하기 쉬운 코드 작성을 지향하도록 하겠습니다..ㅠㅠ!!

hogang1223 replied:
+ 궁금한게 컨벤션에 맞게 짠다는게 무슨 의미인가요....???
제가 또... 강좌 제대로 안읽어서 이해를 못하고 있는걸까요...?
그런거면 다시 읽어보라고 말씀해주세여ㅠ

Kioding replied:
`캠핑사이트 > contents > Week1 > 08. 변수와 메모리 그리고 이름짓기` 에서 나왔던 [Swift API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/) 의 내용입니다 👻

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,92 @@
+    if (topping == nil && body.hasPrefix("*")){
+        return " " + body
+    }else if (topping == nil){
+        return peperoBar + "\n " + body
```
`peperoBar` 가 빼빼로 사이에 들어가 있어 어색하네요. 같은 값이라면 변수의 이름에 상관없이 쓰는 게 좋을까요??

hogang1223 replied:
변수 많이 사용 안하는게 좋다고 생각해서 같은 값이면 그냥 사용하던 편이었습니다만...
이 경우에는 그럼 따로 변수를 선언하거나 혹은 값을 바로 넣어주는게 더 보기 좋을까요?

Kioding replied:
변수를 많이 사용하지 않는 건 좋은 방법인데요!
하지만, 여기서는 `peperoBar` 라는 변수의 네이밍이 코드를 이해하는데 방해요소가 아닐까 싶네요👀
쓰임이 다르다면 새로운 변수 지정 혹은 값을 넣는 게 가독성 측면에서 좋을 것 같습니다 😄

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,92 @@
+func makePeperoBody(body:String, topping:String?) -> String{
+    
+    if (topping == nil && body.hasPrefix("*")){
+        return " " + body
```
이 부분만 한줄 출력을 하고 있는데 이유가 있을까요?

hogang1223 replied:
ㅎㅎ 한줄로 충분히 반복할 수 있는 모양새의 빼빼로라
따로 다른 친구들과 통일시킬 생각을 하지 않았습니다...

역시 두 줄 출력해서 다른 타입의 빼빼로와도 통일감 있게 해주는게 좋을까요?

Kioding replied:
DM으로 코멘트 드렸습니다 😆

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,92 @@
+    
+    for _ in 1...tempLength {
+        print(peperoBody)
+    }
```
많이 고민하신 게 코드에서 엿보이네요👍

위에서 코멘트 드린 것처럼 두줄 출력을 통일화 한다면 for문만으로도 충분할 것 같은데 어떻게 바꿔볼 수 있을까요?

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,92 @@
+}
+
+// 요구사항 1-4 : 각종 매개변수를 받아 위의 세 메서드를 호출하는 함수
+func printPepero(type:PeperoType){
```
- `print` 라는 함수명은 콘솔 출력을 위한 개발자 입장의 함수명인데 다른 사람이 이해하기 좋은 네이밍은 어떤게 있을까요?
- `(type:PeperoType){` 을 붙인 이유가 있을까요?

hogang1223 replied:
- 다른사람이 이해하기 좋은 네이밍....
좀 더 직관적으로 빼빼로를 그린다 라고 표현하자면
drawPepero일까요?
- 하핫...... ㅎ 죄삼다 ㅎ 급하게 치다가 입니다... 별 의미 없습니다.. ㅎㅎ

hogang1223 committed:
fix : 빼빼로 몸통 그리기 함수 수정

- makePeperoBody : lineBreak변수 선언, 모든 빼빼로 몸통 2줄 반환
- printPeperoBody : makePeperoBody 함수 수정에 따른 수정

hogang1223 committed:
fix : printPepero 함수 수정

- 빼빼로 출력 함수 수정 printPepero -> drawPepero

Kioding commented on CodeStarterCamp_Week1/Step2.swift:
```diff
@@ -0,0 +1,67 @@
+        print("")
+    }
+}
+
```
따로 파일을 분리한 점 좋네요!
이제 12~19 줄이 필요없다면 지워도 괜찮겠죠?!

Kioding replied:
이 부분도 컨벤션에 맞게 고쳐보면 좋을 것 같네요!

hogang1223 committed:
style : 매개변수 줄바꿈 수정

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,102 @@
+
+enum PeperoType {
+    case one, two, three, four
+}
```
열거형을 사용하지 않는다면 지워도 되겠네요 :)

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,102 @@
+    print("")
+        
+}
+ */
```
이 부분도 필요없다면 지워도 괜찮겠죠? 👀

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,102 @@
+    let lineBreak = "\n "
+    if (topping == nil && body.hasPrefix("*")){
+        return " " + body + lineBreak + body
+    }else if (topping == nil){
```
- `}else`, `nil){` 이 부분도 조정이 필요해보이네요 :)
- `(topping == nil)` 위에서 말한 것처럼 swift에서는 소괄호가 생략이 가능합니다. 
조건이 하나라면 생략하는 게 가독성에 좋습니다 👍

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,102 @@
+}
+
+// 요구사항 1-4 : 각종 매개변수를 받아 위의 세 메서드를 호출하는 함수
+func drawPepero(bodyShape: String, topping:String?, bodyLength: Int, barLength: Int) {
```
`(bodyShape: String, topping:String?, bodyLength: Int, barLength: Int)` 매개변수와 전달인자의 띄어쓰기가 다른데 통일성을 가져볼까요? 이 함수 뿐 아니라 다른 함수도 한번씩 확인해주세요!

hogang1223 replied:
아이고... 이거 체크한다고 체크했는데 또또또.... 싹 다 체크해서 수정하겠습니다...!

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,102 @@
+    }
+}
+
+
```
여기는 2줄이 떨어져 있는 이유가 있을까요?

hogang1223 replied:
제가 그게  보기 편해서 입니다...

보내주신 스타일 가이드에는.. 한줄씩 띄워져 있네요..ㅎ..
이 부분도 수정 및 익숙해지도록 해보겠습니다~

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,102 @@
+// 요구사항 1-2 : 몸통(String), 토핑(String)을 매개변수로 받아 토핑과 몸통을 그리는 함수
+func makePeperoBody(body:String, topping:String?) -> String {
+    
+    let lineBreak = "\n "
```
자주 쓰는 값을 변수로 지정해준 점 보기 좋습니다 👍

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,102 @@
+func makePeperoBody(body:String, topping:String?) -> String {
+    
+    let lineBreak = "\n "
+    if (topping == nil && body.hasPrefix("*")){
```
swift에서는 소괄호가 생략이 가능한데요!
가독성을 위해 소괄호를 사용하고자 한다면 어떤식으로 사용하는 게 좋을까요?

hogang1223 replied:
전 소괄호 쓰는게 더 가독성이 높아서 주로 사용하는 편입니다만...
흠... 소괄호 사이 한칸씩 빈칸을 주는게 좋은가요?? 아니면 저 비교연산자 (&&)을 앞뒤로 한칸씩 빈칸을 더 주는게 좋은가요....?

```swift
// case 1
if ( topping == nil && body.hasPrefix("*") ) {

// case 2
if (topping == nil  &&  body.hasPrefix("*")) {

// case 3
if (topping == nil) && (body.hasPrefix("*")) {
```

흠... 셋 다 안 익숙해서 그런건지 어떤건지 모르겠지만 별루네여.....

Kioding replied:
Swift는 if 뒤에 오는 조건문에는 소괄호가 오지 않는 걸 지향하는데요!
아래를 한번 보면 어떤 뜻인지 알기 편할 것 같네요. 지금 고쳐주신 코드는 깔끔해보입니다 😆
[Swift Style Guide](https://google.github.io/swift/#parentheses)

hogang1223 committed:
style : 컨벤션에 맞추어 코드 수정

- STEP2, STEP3 불필요한 주석 삭제
- STEP2, STEP3 컨벤션에 맞춰 줄바꿈 및 띄어쓰기 수정
- main : 요구사항 3 추가

hogang1223 committed:
fix : 빼빼로 타입 설정

- 빼빼로 타입 제한 : enum 부활
- 요구사항 : 빼빼로 정보 출력

hogang1223 committed:
docs : step3 순서도 수정

hogang1223 committed:
fix : 빼빼로 몸통 만들기 함수 수정

- makePeperoBody 함수 수정

hogang1223 committed:
docs : 순서도 파일명 수정

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,101 @@
+
+// MARK: 빼빼로 그리기 요구사항1
+
+enum PeperoType {
+    case one, two, three, four
+    case custom1, custom2, custom3
+}
```
enum이 다시 살아났군요 😆
custom1, custom2, custom3 에 새로운 값을 넣는게 아니라면, 한줄로 작성해주시면 좋습니다.

Kioding commented on CodeStarterCamp_Week1/Step3.swift:
```diff
@@ -0,0 +1,101 @@
+    var barLength: Int = 0
+    
+    switch type {
+    case .one:
+        bodyLength = 10
+        barLength = 4
+        break
+    case .two:
+        topping = "&"
+        bodyLength = 12
+        barLength = 4
+        break
+    case .three:
+        topping = "#"
+        bodyLength = 12
+        barLength = 6
+        break
+    case .four:
+        bodyShape = "|0|"
+        bodyLength = 6
+        barLength = 4
+        break
+    case .custom1:
+        bodyShape = "000"
+        topping = "%"
+        bodyLength = 10
+        barLength = 4
+        break
+    case .custom2:
+        bodyShape = "*&*"
+        topping = "&"
+        bodyLength = 6
+        barLength = 4
+        break
+    case .custom3:
+        bodyShape = "+++"
+        bodyLength = 8
+        barLength = 4
+        break
+    }
```
제한된 상품의 경우라면 enum의 활용이 좋은데요!
주문할 때, 몸통과 토핑이 커스텀 가능한 아이스크림이라면 어떨까요?
enum을 어떨 때 활용하면 좋을지 한번 고민해보면 좋을 것 같네요 👻
